client/question/services/FeedbackGeneratorService.js

```diff
@@ -62,6 +121,39 @@ tie.factory('FeedbackGeneratorService', [
     var previousRuntimeErrorString = '';
 
     /**
+     * Pseudo-randomly returns an int between min and max (inclusive).
+     *
+     * @param {number} min Minimum number for range.
+     * @param {number} max Maximum number for range.
+     * @returns {number}
+     * @private
+     */
+    var _getRandomInt = function(min, max) {
+      min = Math.ceil(min);
+      max = Math.floor(max);
+      return Math.floor(Math.random() * (max - min)) + min
+    };
+
+    /**
+     * Randomly selects feedback string from an object containing variations
+     * of a given feedback type.
+     *
+     * @param {Object.<string, Array.<string>>} correctnessFeedbackText
+     * @param {string} type Correctnness feedback type.
+     * @returns {string}
+     * @private
+     */
+    var _selectCorrectnessFeedback = function(correctnessFeedbackText, type) {
+      var defaultFeedbackIndex = lastCorrectnessFeedback[type];
```

Comment:
As a general rule, if the only thing we're referring to in lastCorrectnessFeedback is the type, then we should just store the type. (That said, that variable might not be needed any more if we're keeping track of status as described above.)

Reply:
This is used to make sure we don't present the same feedback message twice. There is an index tied to lastCorrectnessFeedback; lastCorrectnessFeedback[type] stores the index of the feedback last given for the specified type. Does that make sense?

Reply:
In that case, how about just storing a lastCorrectnessFeedbackString instead? If I'm understanding correctly, I don't think this needs to be segmented by type...

Reply:
I might not have been precise enough, I am not showing the same feedback two times in a row for a given feedback type. E.g., 
feedback-type-a-1
feedback-type-b-1
feedback-type-a-2 (feedback-type-a-1 not allowed since it was last used)

Does that make sense?

Reply:
Not quite, sorry. I assume a and b refer to distinct states/statuses, and b comes later in the sequence than a. If that's the case, what I don't understand is why we'd go back to feedback of type a? I thought in the state diagram we discussed, the user can only move forward through it.

Reply:
I wanted to implement variation across correctness tests as well.